### PR TITLE
[BACKLOG-41332] Implemented Move/Copy DB Connections and remaining DB…

### PR DESCRIPTION
**Pull request by pdesai16**
… Actions

on right click in the left hand configuration/database node based on project profile rule

### CODE REVIEW

File: ui/src/main/java/org/pentaho/di/ui/spoon/Spoon.java [hunk 1]
```diff
@@ -2795,16 +2800,35 @@ public void delConnection() {
...
+    DatabaseManagementInterface dbManager = null;
+    DatabaseMeta databaseMeta = null;
+    try {
+      dbManager = getDbManager( leveledSelection.getLevel() );
+      databaseMeta = dbManager.getDatabase( leveledSelection.getName() );
+    } catch ( Exception e ) {
+      new ErrorDialog( shell, BaseMessages.getString( PKG, "Spoon.ErrorDialog.Title" ),
+        BaseMessages.getString( PKG, "Unexpected error retrieving Database Connection", databaseMeta.getName() ), e );
+    }
     delegates.db.sqlConnection( databaseMeta );
```
Comment (bconlon-hv) [inline ui/src/main/java/org/pentaho/di/ui/spoon/Spoon.java:2813]: All of these calls to delegates.db should be inside the try block now right?  Some of them are and some aren't.

File: ui/src/main/java/org/pentaho/di/ui/spoon/delegates/SpoonDBDelegate.java [hunk 2]
```diff
@@ -124,29 +131,52 @@ private DatabaseDialog getDatabaseDialog() {
...
+        selectedName = selectedName.trim();
+        // check if the selectedName already exist
+        if ( selectedName.equals( originalName )
+            && databaseMeta.findDatabase( dbManager.getDatabases(), selectedName ) != null ) {
+          databaseMetaCopy.setName( selectedName );
+          DatabaseDialog.showDatabaseExistsDialog( spoon.getShell(), databaseMetaCopy );
+          return;
+        }
+        databaseMetaCopy.setName( selectedName );
+        databaseMetaCopy.setName( selectedName );
```
Comment (abryant-hv) [inline ui/src/main/java/org/pentaho/di/ui/spoon/delegates/SpoonDBDelegate.java:160]: probably don't need this twice. Was once supposed to be DisplayName?

**Review by abryant-hv — APPROVED** [review]

File: ui/src/main/java/org/pentaho/di/ui/spoon/Spoon.java (hunk 1 above)
Comment (pdesai16) [reply ui/src/main/java/org/pentaho/di/ui/spoon/Spoon.java:2813]: I reused the existing code and the SpoonDbDelegate code is not throwing any exception.

Comment (abryant-hv) [reply ui/src/main/java/org/pentaho/di/ui/spoon/Spoon.java:2813]: Right, but if there was an exception in the try block, and the dialog was displayed, we probably do not want to proceed with the call to the delegate.

Comment (abryant-hv) [reply ui/src/main/java/org/pentaho/di/ui/spoon/Spoon.java:2813]: And I see that you copied this from editConnection(), which I messed up in the first place. ;)

File: ui/src/main/java/org/pentaho/di/ui/spoon/Spoon.java [hunk 3]
```diff
@@ -2795,16 +2800,35 @@ public void delConnection() {
...
+      DatabaseManagementInterface dbManager = null;
+      DatabaseMeta databaseMeta = null;
+      try {
+        dbManager = getDbManager( leveledSelection.getLevel() );
+        databaseMeta = dbManager.getDatabase( leveledSelection.getName() );
+      } catch ( Exception e ) {
+        new ErrorDialog( shell, BaseMessages.getString( PKG, "Spoon.ErrorDialog.Title" ),
+          BaseMessages.getString( PKG, "Unexpected error retrieving Database Connection", databaseMeta.getName() ), e );
+      }
       delegates.db.clearDBCache( databaseMeta );
```
[the remark this answers is not in the document]
Comment (pdesai16) [reply ui/src/main/java/org/pentaho/di/ui/spoon/Spoon.java:2832]: Some of existing menu items that are  calling delegate method  like sqlConnection () and this method are not throwing any exception

**Review by niehusa — APPROVED** [review]

File: ui/src/main/java/org/pentaho/di/ui/spoon/Spoon.java (hunk 1 above)
Comment (abryant-hv) [reply ui/src/main/java/org/pentaho/di/ui/spoon/Spoon.java:2813]: I'm actually surprised the compiler isn't complaining that databaseMeta might not have been initialized.

File: ui/src/main/java/org/pentaho/di/ui/spoon/Spoon.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (abryant-hv) [reply ui/src/main/java/org/pentaho/di/ui/spoon/Spoon.java:2832]: right, see above. The risk is not that clearDBCache will throw, the risk is that getDatabase() could throw, and then clearDBCache would still be called, when it shouldn't. 

Alternatively, we could just `return` after showing the ErrorDialog.

File: ui/src/main/java/org/pentaho/di/ui/spoon/Spoon.java [hunk 4]
```diff
@@ -2756,14 +2756,19 @@ public void editConnection() {
...
+    DatabaseMeta databaseMeta = null;
+    try {
+      dbManager = getDbManager( leveledSelection.getLevel() );
+      databaseMeta = dbManager.getDatabase( leveledSelection.getName() );
+    } catch ( Exception e ) {
+      new ErrorDialog( shell, BaseMessages.getString( PKG, "Spoon.ErrorDialog.Title" ),
+        BaseMessages.getString( PKG, "Unexpected error retrieving Database Connection", databaseMeta.getName() ), e );
+    }
+    getLog().logBasic( "Duplicating the connection " + leveledSelection.getName() );
+    delegates.db.dupeConnection( databaseMeta, dbManager );
```
Comment (abryant-hv) [inline ui/src/main/java/org/pentaho/di/ui/spoon/Spoon.java:2770]: please move this one into the try block as well.

File: ui/src/main/java/org/pentaho/di/ui/spoon/Spoon.java [hunk 5]
```diff
@@ -2869,8 +2888,84 @@ public void exploreDatabase() {
...
+    DatabaseManagementInterface dbManager = null;
+    DatabaseMeta databaseMeta = null;
+    try {
+      dbManager = getDbManager( leveledSelection.getLevel() );
+      databaseMeta = dbManager.getDatabase( leveledSelection.getName() );
+    } catch ( Exception e ) {
+      new ErrorDialog( shell, BaseMessages.getString( PKG, "Spoon.ErrorDialog.Title" ),
+        BaseMessages.getString( PKG, "Unexpected error retrieving Database Connection", databaseMeta.getName() ), e );
+    }
+    delegates.db.exploreDB( databaseMeta, dbManager, true );
```
Comment (abryant-hv) [inline ui/src/main/java/org/pentaho/di/ui/spoon/Spoon.java:2901]: please move this one into the try block as well.

File: ui/src/main/java/org/pentaho/di/ui/spoon/Spoon.java [hunk 6]
```diff
@@ -2756,14 +2756,19 @@ public void editConnection() {
   }
```
Comment (abryant-hv) [inline ui/src/main/java/org/pentaho/di/ui/spoon/Spoon.java:2756]: please move the one here in editConnection() into the try block as well.

**Review by bconlon-hv — APPROVED** [review]

**Pull request merged**